src/cc/BPF.cc

```diff
@@ -0,0 +1,236 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdio>
+#include <exception>
+#include <regex>
+#include <vector>
+
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "compat/linux/bpf.h"
+#include "libbpf.h"
+#include "perf_reader.h"
+
+#include "BPF.h"
+
+namespace ebpf {
+  const static std::regex KPROBE_EV_NAME_REPLACE("[+.]");
+  const static std::regex UPROBE_EV_PATH_REPLACE("[^a-zA-Z0-9_]");
+
+  BPF::~BPF() {
+    for (auto it : kprobes_)
+      try {
+        detach_kprobe_event(it.first, it.second);
+      } catch (const std::exception& e) {
+        fprintf(
+          stderr,
+          "Failed to detach kprobe event %s reader %p in BPF destructor\n",
+          it.first.c_str(), it.second
+        );
+      }
+    for (auto it : uprobes_)
+      try {
+        detach_uprobe_event(it.first, it.second);
+      } catch (const std::exception& e) {
+        fprintf(
+          stderr,
+          "Failed to detach uprobe event %s reader %p in BPF destructor\n",
+          it.first.c_str(), it.second
+        );
+      }
+  }
+
+  void BPF::attach_kprobe(
+    const std::string& kernel_func,
+    const std::string& probe_func,
+    bool ret_probe,
+    pid_t pid, int cpu, int group_fd,
+    perf_reader_cb cb, void* cb_cookie
+  ) {
+    int probe_fd = load_func(probe_func, BPF_PROG_TYPE_KPROBE);
+    if (probe_fd == -1)
+      throw std::runtime_error("Unable to load probe function " + probe_func);
```

Comment:
Will do exception handling and clean up all over the place. Thanks for pointing out.